### ISSUE TITLE
Analysis_pipeline_colab_compatible

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#2]
"""
Reproducible analysis pipeline (Colab-compatible, but also runnable locally).

IMPORTANT:
- Originally written in Japanese; this public version is translated to English
  WITHOUT changing any computational logic.
- The default input path points to a Google Colab + Google Drive environment:
    /content/drive/MyDrive/CommmedpatientTT.xlsx
  Readers MUST replace this path with their own file location.

Reproducibility policy:
- No steps, parameters, or assumptions are omitted.
- All fixed parameters (filters, imputation, scaling, KMeans settings, bins) are explicitly specified.
"""

import os
import zipfile
import argparse
import numpy as np
import pandas as pd

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA

from scipy.stats import kruskal, chi2_contingency
from scipy.cluster.hierarchy import linkage, dendrogram
import matplotlib.pyplot as plt

# -------------------------
# 0) Column specification (19 variables)
# -------------------------
TARGET_COLS = [
    "age", "height", "BW", "BMI", "Hb", "Hct", "Alb", "AST", "ALT", "ALP", "Cre",
    "Glu", "TG", "HDL-C", "LDL-C", "LH", "FSH", "CRP", "TT"
]

# Column-name normalization to absorb common variants (extend if needed)
RENAME_MAP = {
    "HDL": "HDL-C",
    "LDL": "LDL-C",
    "HDL_C": "HDL-C",
    "LDL_C": "LDL-C",
    "BodyWeight": "BW",
    "Weight": "BW",
}

# -------------------------
# Helpers
# -------------------------
def ensure_numeric(df: pd.DataFrame, cols):
    for c in cols:
        df[c] = pd.to_numeric(df[c], errors="coerce")
    df = df.replace([np.inf, -np.inf], np.nan)
    return df

def med_iqr(series: pd.Series) -> str:
    s = pd.to_numeric(series, errors="coerce").dropna()
    if len(s) == 0:
        return ""
    q1 = s.quantile(0.25)
    q3 = s.quantile(0.75)
    med = s.median()
    return f"{med:.2f} ({q1:.2f}–{q3:.2f})"

def age_dist_text(df: pd.DataFrame, labels_age):
    total = df.shape[0]
    out = {}
    for g in labels_age:
        n = int((df["age_group"] == g).sum())
        pct = (n / total * 100) if total > 0 else 0.0
        out[g] = f"{n} ({pct:.1f}%)"
    return out

def safe_colab_download(path: str):
    """
    Download only when running inside Google Colab.
    This keeps the SAME outputs but avoids crashing outside Colab.
    """
    try:
        from google.colab import files  # type: ignore
        files.download(path)
    except Exception:
        # Not in Colab or download not available; do nothing.
        pass

# -------------------------
# Main
# -------------------------
def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--xlsx_path",
        type=str,
        default="/content/drive/MyDrive/CommmedpatientTT.xlsx",
        help=(
            "Path to the input Excel file. "
            "Default is a Colab+Google Drive example; replace with your own path."
        ),
    )
    parser.add_argument(
        "--outdir",
        type=str,
        default="/content/e1_outputs",
        help="Output directory for E-1 style outputs (CSV/figures).",
    )
    parser.add_argument(
        "--table1_outdir",
        type=str,
        default="/content/table1_K3_outputs",
        help="Output directory for Table 1 (K=3) package (CSV/ZIP/meta).",
    )
    args = parser.parse_args()

    # -------------------------
    # 1) Load
    # -------------------------
    if not os.path.exists(args.xlsx_path):
        raise FileNotFoundError(f"File not found: {args.xlsx_path}")

    os.makedirs(args.outdir, exist_ok=True)

    df = pd.read_excel(args.xlsx_path)
    print("Loaded:", df.shape)
    print("Columns:", df.columns.tolist())

    # Normalize column names (if present)
    df = df.rename(columns={k: v for k, v in RENAME_MAP.items() if k in df.columns})

    missing_cols = [c for c in TARGET_COLS if c not in df.columns]
    if missing_cols:
        raise ValueError(f"Missing required columns: {missing_cols}")

    X_raw = df[TARGET_COLS].copy()
    print("X_raw:", X_raw.shape)
    print("NA counts (top 10):")
    print(X_raw.isna().sum().sort_values(ascending=False).head(10))

    # -------------------------
    # 2) Subjects filter: age & BMI complete (strict)
    # -------------------------
    n_before = len(X_raw)
    n_missing_age = int(X_raw["age"].isna().sum())
    n_missing_bmi = int(X_raw["BMI"].isna().sum())
    print("Missing age:", n_missing_age)
    print("Missing BMI:", n_missing_bmi)

    mask_complete_age_bmi = X_raw["age"].notna() & X_raw["BMI"].notna()
    X_raw2 = X_raw.loc[mask_complete_age_bmi].copy()

    print("Rows before:", n_before)
    print("Rows after age/BMI filter:", len(X_raw2))

    # Ensure numeric and finite handling (as in your dendrogram block)
    X_raw2 = ensure_numeric(X_raw2, TARGET_COLS)

    # -------------------------
    # 3) Mean imputation -> standardization (full 19 vars)
    # -------------------------
    imputer = SimpleImputer(strategy="mean")
    scaler = StandardScaler()

    X_imp = imputer.fit_transform(X_raw2.values)
    X_std = scaler.fit_transform(X_imp)

    print("X_std shape:", X_std.shape)  # expected (N, 19)

    # -------------------------
    # 4) PCA explained variance (report only; deterministic)
    # -------------------------
    pca = PCA(n_components=X_std.shape[1], random_state=0)
    pca.fit(X_std)

    evr = pca.explained_variance_ratio_
    pc1 = evr[0] * 100
    pc2 = evr[1] * 100
    pc12 = (evr[0] + evr[1]) * 100

    print(f"PC1 (%): {pc1:.2f}")
    print(f"PC2 (%): {pc2:.2f}")
    print(f"PC1+PC2 (%): {pc12:.2f}")

    cum = np.cumsum(evr) * 100
    print("\nTop 10 cumulative (%):")
    for i in range(min(10, len(cum))):
        print(f"PC1..PC{i+1}: {cum[i]:.2f}")

    # Save PCA variance to CSV (fully reproducible artifact)
    pca_csv = os.path.join(args.outdir, "pca_explained_variance_ratio.csv")
    pd.DataFrame({
        "PC": [f"PC{i+1}" for i in range(len(evr))],
        "explained_variance_ratio": evr,
        "explained_variance_ratio_percent": evr * 100,
        "cumulative_percent": cum,
    }).to_csv(pca_csv, index=False)
    print("Saved:", pca_csv)

    # -------------------------
    # 5) Table 1 generator (K=3) — EXACT as your logic, but de-duplicated and English
    # -------------------------
    K = 3
    random_state = 0
    n_init = 50

    km = KMeans(n_clusters=K, random_state=random_state, n_init=n_init)
    labels = km.fit_predict(X_std)

    X_subj = X_raw2.copy()
    X_subj["cluster"] = labels

    print("\n[KMeans]")
    print("K:", K, "random_state:", random_state, "n_init:", n_init)
    print("Cluster sizes:")
    print(X_subj["cluster"].value_counts().sort_index())

    # Continuous variables list (exactly the 19 vars)
    numeric_vars = TARGET_COLS[:]  # fixed 19 variables

    # Build Table 1: columns = All + each cluster + P value
    N = int(X_subj.shape[0])
    col_all = f"All (n = {N:,})"
    cluster_cols = []
    for c in sorted(X_subj["cluster"].unique()):
        cluster_cols.append(f"Cluster {c} (n = {int((X_subj['cluster']==c).sum()):,})")

    table = pd.DataFrame(index=numeric_vars, columns=[col_all] + cluster_cols + ["P value"])

    # Fill All + clusters (median [IQR])
    for v in numeric_vars:
        table.loc[v, col_all] = med_iqr(X_subj[v])
        for c in sorted(X_subj["cluster"].unique()):
            table.loc[v, f"Cluster {c} (n = {int((X_subj['cluster']==c).sum()):,})"] = med_iqr(
                X_subj.loc[X_subj["cluster"] == c, v]
            )

    # P values: Kruskal–Wallis across clusters (continuous)
    for v in numeric_vars:
        groups = []
        ok = True
        for c in sorted(X_subj["cluster"].unique()):
            g = pd.to_numeric(X_subj.loc[X_subj["cluster"] == c, v], errors="coerce").dropna().values
            if len(g) < 2:
                ok = False
            groups.append(g)
        if ok:
            try:
                p = kruskal(*groups).pvalue
                table.loc[v, "P value"] = f"{p:.3g}" if p >= 0.001 else "<0.001"
            except Exception:
                table.loc[v, "P value"] = ""
        else:
            table.loc[v, "P value"] = ""

    # Age-group distribution rows (exact bins/labels from your code)
    bins = [60, 70, 80, 90, np.inf]
    labels_age = ["60–69", "70–79", "80–89", "90+"]
    X_subj["age_group"] = pd.cut(X_subj["age"], bins=bins, right=False, labels=labels_age)

    age_header = "Proportion of individuals by age group (n, %)"
    insert_rows = [age_header] + labels_age

    idx = list(table.index)
    pos = idx.index("age") + 1 if "age" in idx else 0
    new_index = idx[:pos] + insert_rows + idx[pos:]
    table = table.reindex(new_index)

    table.loc[age_header, :] = ""

    dist_all = age_dist_text(X_subj, labels_age)
    for g in labels_age:
        table.loc[g, col_all] = dist_all[g]
        for c in sorted(X_subj["cluster"].unique()):
            dfc = X_subj.loc[X_subj["cluster"] == c]
            table.loc[g, f"Cluster {c} (n = {int((X_subj['cluster']==c).sum()):,})"] = age_dist_text(dfc, labels_age)[g]

    # Chi-square P value for age-group distribution
    cont = pd.crosstab(X_subj["cluster"], X_subj["age_group"])
    cont = cont.reindex(columns=labels_age, fill_value=0)
    try:
        chi2, p_age, dof, exp = chi2_contingency(cont.values)
        ptxt = f"{p_age:.3g}" if p_age >= 0.001 else "<0.001"
    except Exception:
        ptxt = ""
    table.loc[age_header, "P value"] = ptxt

    # Optional: rename index (your mapping) — kept exactly
    tidy_map = {
        "age": "Age (years)",
        "height": "Height (cm)",
        "BW": "Body weight (kg)",
        "BMI": "BMI (kg/m²)",
        "Hb": "Hemoglobin (g/dL)",
        "Hct": "Hematocrit (%)",
        "Alb": "Albumin (g/dL)",
        "AST": "AST (U/L)",
        "ALT": "ALT (U/L)",
        "ALP": "ALP (U/L)",
        "Cre": "Creatinine (mg/dL)",
        "Glu": "Glucose (mg/dL)",
        "TG": "Triglycerides (mg/dL)",
        "HDL-C": "HDL (mg/dL)",
        "LDL-C": "LDL (mg/dL)",
        "LH": "LH (mIU/mL)",
        "FSH": "FSH (mIU/mL)",
        "CRP": "CRP (mg/dL)",
        "TT": "Total testosterone (ng/mL)",
    }
    table.index = [tidy_map.get(x, x) for x in table.index]

    # Save outputs + ZIP + (Colab) download
    os.makedirs(args.table1_outdir, exist_ok=True)

    table_path = os.path.join(args.table1_outdir, "Table1_K3_median_IQR_with_agegroups.csv")
    table.to_csv(table_path, encoding="utf-8-sig")

    sizes_path = os.path.join(args.table1_outdir, "cluster_sizes_K3.csv")
    X_subj["cluster"].value_counts().sort_index().to_csv(sizes_path, header=["count"])

    meta_path = os.path.join(args.table1_outdir, "table1_K3_meta.txt")
    with open(meta_path, "w") as f:
        f.write("Reproducibility metadata (no omissions)\n")
        f.write(f"Input file: {args.xlsx_path}\n")
        f.write(f"N(age & BMI complete) = {N}\n")
        f.write("Preprocessing: mean imputation for remaining vars -> standardization\n")
        f.write(f"KMeans: K={K}, random_state={random_state}, n_init={n_init}\n")
        f.write("Age-group bins: [60,70,80,90,inf), right=False\n")
        f.write("Age-group labels: 60–69 / 70–79 / 80–89 / 90+\n")
        f.write("Cluster sizes:\n")
        f.write(X_subj["cluster"].value_counts().sort_index().to_string())
        f.write("\n")

    zip_path = os.path.join(args.table1_outdir, "table1_K3_outputs.zip")
    with zipfile.ZipFile(zip_path, "w", zipfile.ZIP_DEFLATED) as z:
        z.write(table_path, arcname=os.path.basename(table_path))
        z.write(sizes_path, arcname=os.path.basename(sizes_path))
        z.write(meta_path, arcname=os.path.basename(meta_path))

    print("\n[Table1 outputs]")
    print("Saved:", table_path)
    print("Saved:", sizes_path)
    print("Saved:", meta_path)
    print("ZIP:", zip_path)

    safe_colab_download(zip_path)

    # -------------------------
    # 6) Dendrogram (Ward) — exploratory visualization (same preprocessing)
    # -------------------------
    Z = linkage(X_std, method="ward")  # Ward assumes Euclidean distance

    plt.figure(figsize=(10, 6))
    dendrogram(
        Z,
        no_labels=True,
        color_threshold=None,
    )
    plt.xlabel("Subjects")
    plt.ylabel("Euclidean distance (Ward linkage)")
    plt.title("Hierarchical clustering dendrogram (exploratory visualization)")
    plt.tight_layout()

    dendro_png = os.path.join(args.outdir, "dendrogram_ward_exploratory.png")
    plt.savefig(dendro_png, dpi=300)
    plt.show()
    print("Saved:", dendro_png)

if __name__ == "__main__":
    main()
